From ff75a12f886812c587d0c5a01093189f2a84ada0 Mon Sep 17 00:00:00 2001
From: Wang ShaoBo <bobo.shaobowang@huawei.com>
Date: Tue, 5 Jul 2022 19:38:35 +0800
Subject: [PATCH 3299/4585] [Huawei] arm64/mpam: make mbw_max/min not less than
 min_bw

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187154

--------------------------------

Do not allow min_bw below the granularity with adjusting mbw_max/min,
and with setting mbw_max/min less than min_bw, return 'Invalid argument'
directly.

Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
---
 arch/arm64/kernel/mpam/mpam_resctrl.c | 8 ++------
 arch/arm64/kernel/mpam/mpam_setup.c   | 3 +++
 2 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/arch/arm64/kernel/mpam/mpam_resctrl.c b/arch/arm64/kernel/mpam/mpam_resctrl.c
index d00746c08922..f9a360a4c718 100644
--- a/arch/arm64/kernel/mpam/mpam_resctrl.c
+++ b/arch/arm64/kernel/mpam/mpam_resctrl.c
@@ -338,15 +338,11 @@ parse_bw(char *buf, struct resctrl_resource *r,
 	switch (rr->ctrl_features[type].evt) {
 	case QOS_MBA_MAX_EVENT_ID:
 	case QOS_MBA_PBM_EVENT_ID:
-		if (kstrtoul(buf, rr->ctrl_features[type].base, &data))
-			return -EINVAL;
-		data = (data < r->mbw.min_bw) ? r->mbw.min_bw : data;
-		data = roundup(data, r->mbw.bw_gran);
-		break;
 	case QOS_MBA_MIN_EVENT_ID:
 		if (kstrtoul(buf, rr->ctrl_features[type].base, &data))
 			return -EINVAL;
-		/* for mbw min feature, 0 of setting is allowed */
+		if (data < r->mbw.min_bw)
+			return -EINVAL;
 		data = roundup(data, r->mbw.bw_gran);
 		break;
 	default:
diff --git a/arch/arm64/kernel/mpam/mpam_setup.c b/arch/arm64/kernel/mpam/mpam_setup.c
index 6e71c99d19b0..cef66a5e6f2f 100644
--- a/arch/arm64/kernel/mpam/mpam_setup.c
+++ b/arch/arm64/kernel/mpam/mpam_setup.c
@@ -419,6 +419,9 @@ static int mpam_resctrl_resource_init(struct mpam_resctrl_res *res)
 		 * of 1 would appear too fine to make percentage conversions.
 		 */
 		r->mbw.bw_gran = GRAN_MBA_BW;
+		/* do not allow mbw_max/min below mbw.bw_gran */
+		if (r->mbw.min_bw < r->mbw.bw_gran)
+			r->mbw.min_bw = r->mbw.bw_gran;
 
 		/* We will only pick a class that can monitor and control */
 		r->alloc_capable = true;
-- 
2.17.1

